### Optimal Play v2.0

In this document, I will attempt to create an agent that plays optimally. My defintion of optimal is the following:
<br>

$$\text{Always pick the move with the highest probability of achieving a win state}$$
<br>

In code, I want a function that inputs the gamestate and roll and outputs the move with the highest probability of achieving a win state.

#### Outline 
<br>

1) Uniquely ID every possible gamestate such that the ID can be used both as a unique pointer to a location and a gamestate
2) Create a list of all possible winstates for white and black 
3) Create a dictionary of 
end) A function that inputs gamestate and roll and outputs the move with the highest probability of achieving a win state

In [2]:
import numpy as np
import time

In [3]:
current_game_state = np.array([[(0,0), (1,0), (1,0), (0,0)], 
                               [(0,1), (0,1), (0,1), (0,1)], 
                               [(0,0), (0,1), (0,1), (0,0)]], 
                               dtype = [('whites', 'i4'), ('blacks', 'i4')])

ID unique ordering
<br> 
W2, W3, N1, N2, N3, N4, B2, B3, W1, W4, B1, B4

In [4]:
# Function 0: Convert an array into a unique string
def array_to_ID(array):
    '''
    Takes a 3x8 array and converts it into a unique ID 
    '''
    reshaped_array = array.reshape(1,12)
    ID = ''
    square_0 = tuple(reshaped_array[0][0])
    square_1 = tuple(reshaped_array[0][1])
    square_2 = tuple(reshaped_array[0][2])
    square_3 = tuple(reshaped_array[0][3])
    square_4 = tuple(reshaped_array[0][4])
    square_5 = tuple(reshaped_array[0][5])
    square_6 = tuple(reshaped_array[0][6])
    square_7 = tuple(reshaped_array[0][7])
    square_8 = tuple(reshaped_array[0][8])
    square_9 = tuple(reshaped_array[0][9])
    square_10 = tuple(reshaped_array[0][10])
    square_11 = tuple(reshaped_array[0][11])
    if square_0 == (0,0):
        ID += '0'
    if square_0 == (1,0):
        ID += '1'
    if square_3 == (0,0):
        ID += '0'
    if square_3 == (1,0):
        ID += '1'
    if square_4 == (0,0):
        ID += '0'
    if square_4 == (1,0):
        ID += '1'
    if square_4 == (0,1):
        ID += '2'
    if square_5 == (0,0):
        ID += '0'
    if square_5 == (1,0):
        ID += '1'
    if square_5 == (0,1):
        ID += '2'
    if square_6 == (0,0):
        ID += '0'
    if square_6 == (1,0):
        ID += '1'
    if square_6 == (0,1):
        ID += '2'
    if square_7 == (0,0):
        ID += '0'
    if square_7 == (1,0):
        ID += '1'
    if square_7 == (0,1):
        ID += '2'
    if square_8 == (0,0):
        ID += '0'
    if square_8 == (0,1):
        ID += '2'
    if square_11 == (0,0):
        ID += '0'
    if square_11 == (0,1):
        ID += '2'
    if square_1 == (0,0):
        ID += '0'
    if square_1 == (1,0):
        ID += '1'
    if square_1 == (2,0):
        ID += '3'
    if square_2 == (0,0):
        ID += '0'
    if square_2 == (1,0):
        ID += '1'
    if square_2 == (2,0):
        ID += '3'
    if square_9 == (0,0):
        ID += '0'
    if square_9 == (0,1):
        ID += '2'
    if square_9 == (0,2):
        ID += '4'
    if square_10 == (0,0):
        ID += '0'
    if square_10 == (0,1):
        ID += '2'
    if square_10 == (0,2):
        ID += '4'
    
    return ID

In [16]:
ID = array_to_ID(current_game_state)
print(ID)
print(int(ID))

002222001122
2222001122


In [17]:
# Function 2: Convert a str into a tuple 
def num_to_tuple(number):
    '''
    Converts a real number into a tuple 
    '''
    if number == '0':
        return (0,0)
    if number == '1':
        return (1,0)
    if number == '2': 
        return (0,1)
    if number == '3':
        return (2,0)
    if number == '4':
        return (0,2)
    return 

In [18]:
# Function 3: Convert the unique ID back into the numpy array 
def unique_to_array(ID):
    '''
    
    '''
    s = np.array([[num_to_tuple(ID[0]), num_to_tuple(ID[8]), num_to_tuple(ID[9]), 
                   num_to_tuple(ID[1]), num_to_tuple(ID[2]), num_to_tuple(ID[3]), 
                   num_to_tuple(ID[4]), num_to_tuple(ID[5]), num_to_tuple(ID[6]), 
                   num_to_tuple(ID[10]), num_to_tuple(ID[11]), num_to_tuple(ID[7])]], 
                               dtype = [('whites', 'i4'), ('blacks', 'i4')])
    s = s.reshape(3,4)
    
    return s

In [19]:

gs = unique_to_array(ID)
print(gs)

[[(0, 0) (1, 0) (1, 0) (0, 0)]
 [(0, 1) (0, 1) (0, 1) (0, 1)]
 [(0, 0) (0, 1) (0, 1) (0, 0)]]


In [20]:
import string
digs = string.digits + string.ascii_letters
def int2base1(x, base):
    if x == 0:
        return '0000000' + ''.join(digs[0])

    digits = []

    while x:
        digits.append(digs[int(x % base)])
        x = int(x / base)

    digits.reverse()
    
    
    if len(digits) == 1:
        y = '0000000'+ ''.join(digits)   
    if len(digits) == 2:
        y = '000000'+ ''.join(digits)
    if len(digits) == 3:
        y = '00000'+ ''.join(digits)
    if len(digits) == 4:
        y = '0000'+ ''.join(digits)
    if len(digits) == 5:
        y = '000'+ ''.join(digits)
    if len(digits) == 6:
        y = '00'+ ''.join(digits)
    if len(digits) == 7:
        y = '0' + ''.join(digits)
    if len(digits) == 8:
        y = ''.join(digits)

    return y
    

In [21]:
import string
digs = string.digits + string.ascii_letters
def int2base2(x, base):
    if x == 0:
        return '000' + ''.join(digs[0])

    digits = []

    while x:
        digits.append(digs[int(x % base)])
        x = int(x / base)

    digits.reverse()
    
    
    if len(digits) == 1:
        y = '000'+ ''.join(digits)   
    if len(digits) == 2:
        y = '00'+ ''.join(digits)
    if len(digits) == 3:
        y = '0'+ ''.join(digits)
    if len(digits) == 4:
        y = ''+ ''.join(digits)


    return y
    

In [22]:
print(int('11220000', 3))
print(int2base(2187, 3))
print(int('3040', 5))

3564


NameError: name 'int2base' is not defined

In [23]:
# Function 4: Calculate all possible gamestates 
def calculate_states():
    '''
    
    '''
    gamestates = [] 
    
    for i in range(0,3564+1):
        ID = str(int2base1(i, 3))
        for j in range(0, 395+1):
            ID = str(int2base1(i, 3))
            ID += str(int2base2(j, 5))
            gamestates.append(ID)
        
    return gamestates

In [24]:
time_start = time.time()
game_states = calculate_states()
time_end = time.time()
time_taken = time_end - time_start 
print(time_taken, "seconds")
print(len(game_states))

9.811745166778564 seconds
1411740


In [25]:
print(game_states[1411739])
import sys
sys.getsizeof(game_states)

112200003040


12383856

We now have a list that contains at least all of the possible gamestates in our game of ur. It contains 1,411,740 elements. Given any of the strings, we can reconstruct a gamestate, remembering that some of them are valid and some are not. 

What's next? 

Create a mapping from one state to a previous state 

In [26]:
# Function 5: create two lists of all black win states and white win states

def gs_to_wgs(gs):
    '''
    gs = list of gamestates
    '''
    
    white_wgs = []
    black_wgs = []
    
    for i in gs:
        if i[9] == '3' and i[8] == '0' and i[11] != '4' and i[0] == '0' and i[1] == '0' and i[4] != '1' and i[5] != '1' and i[6] != '1' and i[7] != '1':
            white_wgs.append(i)
        if i[11] == '4' and i[10] == '0': # black has two pieces in the end square and home must have 0
            if i[9] != '3' and i[8] != '4' and i[8] != '2' and i[9] != '4' and i[9] != '2': # white cannot have two pieces in his end square and white's home/end cannot have one or two black pieces
                if i[1] != '2' and i[2] != '2' and i[3] != '2' and i[4] != '2' and i[5] != '2' and i[6] != '2' and i[7] != '2': # there can be no black pieces on the board
                    if (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) + int(i[6]) + int(i[7])) <= 2: #The sum of the first 0-7 digits must be less than or equal to 2
                        if i[6] != '1' and i[7] != '1': # cannot be white pieces in black only tiles
                            if i[8] == '3' and (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) > 0): #if 2 whites in white's home, there must be no whites on board
                                continue
                            if i[8] == '1' and i[9] == '1' and (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) > 0): #if there is one white in end and last tile, sum of beginning should be 0 
                                continue
                            if i[8] == '1' and i[9] == '0' and (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) != 1): #if there is one white in end and last tile, sum of beginning should be 0 
                                continue
                            if i[8] == '0' and i[9] == '1' and (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) != 1): #if there is one white in end and last tile, sum of beginning should be 0 
                                continue
                            if i[8] == '0' and i[9] == '0' and (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) != 2): #if there is one white in end and last tile, sum of beginning should be 0 
                                continue
                                
                                
                            black_wgs.append(i) 
    
    return white_wgs, black_wgs

In [27]:
white_wgs, black_wgs = gs_to_wgs(game_states)

In [28]:
print(len(white_wgs))
print(len(black_wgs))
print()

2880
29



            if i[8] != '2' and i[10] =='0' and i[9] != '3' and i[9] != '2' and i[9] != '4' and i[1] != '2' and i[2] != '2' and i[3] != '2' and i[4] != '2' and i[5] != '2' and i[6] == '0' and i[7] == '0':
            if i[8] == '3' and (int(i[0]) + int(i[1]) + int(i[2]) + int(i[3]) + int(i[4]) + int(i[5]) + int(i[6]) + int(i[7])) != 0:
                break

In [29]:
for i in black_wgs:
    print(i)
print(len(black_wgs))

000000001104
000000003004
000001000104
000001001004
000010000104
000010001004
000011000004
000100000104
000100001004
000101000004
000110000004
001000000104
001000001004
001001000004
001010000004
001100000004
010000000104
010000001004
010001000004
010010000004
010100000004
011000000004
100000000104
100000001004
100001000004
100010000004
100100000004
101000000004
110000000004
29


In order to get the win probability from a given game state, you can simulate the game some number of times in order to see how many times that player will win. 

### Variables

TW = dictionary of dictionaries for white
<br>
TB = dictionary of dictionaries for black 
<br>
W_1 = TW['1'] = dictionary for white with roll 1 
<br>
W_2 = TW['2'] = dictionary for white with roll 2 
<br>
B_1 = TB['1'] = dictionary for black with roll 1 
<br>
B_2 = TB['2'] = dictionary for black with roll 2
<br> 
Example of a dictionary entry 
<br>
W_1['030400000000'] = ('110400000000', '000002000302', 0.12)

The first entry is the optimal move, the second entry is the endstate that is most likely from the current gamestate, and the third entry is the probability of getting to endstate given in the second entry.

The next step is calculate which moves are optimal, which endstate is most likely from a given state, and the probability of getting to that end state from the current endstate. Plan of attack is to write three helper functions that do those three tasks and write one function that builds the dictionary of dictionaries. 
<br>


### Dynamic Programming

1) Specify the gamestates that are roll=1 and roll=2 away from a win state for both black and white. Useful: there are 29 winstates for black and 29 winstates for white. There are 683 possible game states. 

In [ ]:
def one_or_two_away():
    '''
    '''
    return 


$\text{[sb]}^2.$